### ISSUE TITLE
consensus: Emergency Mode

### ISSUE THREAD

**Issue by fed-franz** [dusk-network/rusk#1171]
#### Summary

To deal with extreme cases in which the network fails to produce blocks for several iterations, we introduce an Emergency mode, where the step timeouts are relaxed, allowing for time-unbound parallel iterations, and, eventually, an emergency block produced by Dusk.

#### Possible solution design or implementation
Proposed design:
 - Nodes enter Emergency Mode when reaching a very high iteration (e.g., 250)
 - In Emergency Mode, step timeouts are only used to start the following step or iteration, but do not produce a Timeout result. Instead, the steps are kept "active" indefinitely. 
 For instance, if the Generation step timeout expires, the Provisioner participating in the First Reduction will not vote Timeout but wait indefinitely for the candidate block (or a quorum of votes) to be received. 
 In other words, iterations in Emergency Mode only terminates upon reaching consensus on a candidate block. 
 These iterations run in parallel, increasing the chances of a block to be produced and agreed upon.
 
In case not even the emergency mode allows the network to produce a block, he very last iteration (255) is reserved for a Dusk-signed block that is accepted without any quorum. This block will be added to the local chain as ACCEPTED, meaning that it can be replaced by a lower-iteration block, if it exists and reached a quorum.
This mechanism ensures that a block is produced even if the network collapses or gets split.

**Comment by goshawk-3** [comment]
> These iterations run in parallel, increasing the chances of a block to be produced and agreed upon.

@fed-franz Are you suggesting to technically implement concurrent iterations in Emergency Mode? As you know, iterations are now running sequentially. Implementation-wise, it is doable to run a separate `tokio::task` per an iteration  in `Emergency Mode`  but I'd like to ensure if this is worthy.

**Comment by goshawk-3** [comment]
> In case not even the emergency mode allows the network to produce a block, he very last iteration (255) is reserved for a Dusk-signed block that is accepted without any quorum. This block will be added to the local chain as ACCEPTED, meaning that it can be replaced by a lower-iteration block, if it exists and reached a quorum.
This mechanism ensures that a block is produced even if the network collapses or gets split

Implementation-wise, I'd suggest to address this idea with a separate issue. I reckon that we'll need sort of a master key in Genesis state together with `additional` check in `block acceptance procedure` to accept `Dusk-signed block` .

FYI @fed-franz @herr-seppia Shall I file another issue?

**Comment by fed-franz** [comment]
> > These iterations run in parallel, increasing the chances of a block to be produced and agreed upon.
> 
> @fed-franz Are you suggesting to technically implement concurrent iterations in Emergency Mode? As you know, iterations are now running sequentially. Implementation-wise, it might be doable to run a separate `tokio::task` per an iteration  in `Emergency Mode`  but I'd like to ensure if this is worthy.
> 
> 

Yes. The idea is that the last iterations will start like other ones, when the previous iteration timeout expires, but the thread keeps running until all steps are completed (ie, a candidate is received and two quorums are reached). This is akin to what we did when allowing previous iteration messages to be processed (vote for older candidates). This will allow for multiple iterations to reach consensus, at the expense of possible forks.
In particular this mechanism was decided to avoid that reaching the last iteration without a block would completely halt consensus.

Note that we're talking of few iterations, like 5 or 10 tops

**Comment by fed-franz** [comment]
> > In case not even the emergency mode allows the network to produce a block, he very last iteration (255) is reserved for a Dusk-signed block that is accepted without any quorum. This block will be added to the local chain as ACCEPTED, meaning that it can be replaced by a lower-iteration block, if it exists and reached a quorum.
> This mechanism ensures that a block is produced even if the network collapses or gets split
> 
> Implementation-wise, I'd suggest to address this idea with a separate issue. I reckon that we'll need sort of a master key in Genesis state together with `additional` check in `block acceptance procedure` to accept `Dusk-signed block` .
> 
> FYI @fed-franz @herr-seppia Shall I file another issue?

I agree
